### PR TITLE
fix(appsync): sanitized datasource name isn't exported

### PR DESCRIPTION
**Pull request by ramblingenzyme**
Closes https://github.com/aws/aws-cdk/issues/23743
When providing a name with unsupported characters to an AppSync datasource, it is sanitized internally. https://github.com/aws/aws-cdk/pull/22378 introduced a bug where the sanitized name was no longer exposed on DataSource class instances.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Construct Runtime Dependencies:

* [ ] This PR adds new construct runtime dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-construct-runtime-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

image

image

The above code generates the shown diff. This is due to https://github.com/aws/aws-cdk/issues/23743 where the name is sanitised, but not when setting `this.name`.

### CODE REVIEW

**Comment by MrArnoldPalmer** [comment]
hey @ramblingenzyme good catch on this. Are you able to add a unit test here to assert that the sanitized name is correctly returned on `this.name`? An integ test isn't necessary.

**Comment by ramblingenzyme** [comment]
@MrArnoldPalmer Happy to try!

**Comment by ramblingenzyme** [comment]
Now that I'm looking at this, this fix is only going to apply to stuff without tokens, i.e. if a token returns a string with a `-` or there just is one in the static part of the string.

I'm a bit unsure about how AppSync/CF deals with this, and if that situation will just fail to deploy.

**Comment by ramblingenzyme** [comment]
@MrArnoldPalmer I've added a test case for what's definitely fixed by this PR and 2 which aren't covered. I'm unsure if these are possible to be resolved, but are definitely beyond me at this point.

If these are solvable issues, I'm happy to open issues that get linked in the test case comment similar to https://github.com/aws/aws-cdk/blob/74318c7d22bfc00de9e005f68a0a6aaa58c7db39/packages/%40aws-cdk/pipelines/test/docker-credentials.test.ts

**Comment by MrArnoldPalmer** [comment]
So I'm not sure the exact case of having a string that is part token then concatenated with `-` unless the user is interpolating that themselves, in which case the name will not be sanitized and the deploy will fail with an error saying the name is invalid right? Basically I don't think we need to cover this case. Sanitizing is a way to ensure that names originating in the CDK app somewhere wont cause a failure, but if it originates elsewhere there isn't much we can do about it until resolution time. We could make the sanitization itself lazy but I'm not sure how to assure the token is resolved before the sanitization function occurs.

Basically I think we can remove the two failing test cases and merge the fix as is to solve the existing issue.

**Comment by ramblingenzyme** [comment]
Just wasn't sure if that was a normal thing to do, since one of the test cases added in https://github.com/aws/aws-cdk/pull/22378 also has a template string with an embedded token. Removed those test cases and updating the PR message now.

**Review by MrArnoldPalmer — APPROVED** [review]
